src/main/java/com/rackspace/salus/resource_management/web/model/ResourceCreate.java

```
@@ -28,6 +30,7 @@
 public class ResourceCreate implements Serializable {
     @Pattern(regexp="[A-Za-z0-9:-]+")
     @NotBlank
+    @JsonDeserialize(using = JsonToLowerCaseDeserializer.class)
     String resourceId;
```

@jjbuchan @GeorgeJahad, I wanted to gather your thoughts since I'm on the fence about whether the resource ID given by the user should be altered in order to address case insensitive matching via etcd and `Objects.equals`. For example, if a user created a resource with "MyCustomResourceId" then the entity stored and the returned object will turn that into "mycustomresourceid".

---

it seems like if we are only going to allow lower case resourceid's then trying to create a resource like "MyCustomResourceId" should return an error.

---

Without having read the pr yet or thought about all the problems my preference is that we would store (and return) the user provided value but treat it as case-insensitive.

1. the user should always see the same value they provided
1. failures should occur if they try to create mycustomresourceid when MyCustomResourceId already exists
1. other than that create validation our services should treat MyCustomResourceId and mycustomresourceid as the same thing

---

The more I think about it the more I agree with James. 

Per James suggestions this means just changing the collation in the database isn't an option. If we change that to a binary format then we lose case insensitive storage in the db. 

We don't want to do what I have done in this PR in altering the users inputs, which everyone seems to agree on (I didn't feel super comfortable about it but at the time I wasn't positive of a better way to do it)

And to be frank I don't want to go through the entire codebase to make sure all of the comparisons are correct. That is tedious and a recipe for bugs. 

I think that leaves the only option of extending String into a `CaseInsensitiveString` class and overriding the comparison functions to do case insensitive comparisons. This guarantees that comparison will be done correctly whenever it is done on these objects without the chance of introducing bugs and still allows the user input to be stored as is in the database. We can still interact with them as if they are Strings and shouldn't have any storage issues either.

---

Again, I've not read the PR and haven't thought about all the problem... but from what I remember the main point of this issue is that we were already treating it case-insensitive in most apps but etcd was not?  So if we just make etcd less strict (or the interactions with etcd more consistent) maybe that eliminated the problem?

I can look more at this on Monday in case that doesn't make sense.

---

Yeah I think I misunderstood the original intent of the ticket. This makes it a lot more clear to me. I think my idea about extending String is still something that will need to happen, but the other half that I didn't understand was relaxing etcd storage validation to being case insensitive.